torch/csrc/jit/codegen/cuda/executor.cpp

```diff
@@ -185,48 +185,63 @@ namespace {
 
 at::Tensor inferAndAlloc(
     const kir::TensorView* tv,
+    const std::vector<kir::Val*>& sizes,
     kir::ExpressionEvaluator& expr_eval,
     const CompileOptions& options,
     bool zero_init = false) {
   FUSER_PERF_SCOPE("inferAndAlloc");
 
-  std::vector<int64_t> sizes;
+  std::vector<int64_t> inferred_sizes;
 
-  const auto domain = tv->domain();
-  const auto maybe_rfactor_domain =
-      domain->hasRFactor() ? domain->rfactorDomain() : domain->rootDomain();
-
-  for (const auto id : maybe_rfactor_domain) {
-    if (id->isReduction() ||
-        id->iterType() == IterType::BroadcastWithoutStride) {
-      continue;
-    }
-    const auto inferred_val = expr_eval.evaluate(id->rawExtent());
+  for (const auto size : sizes) {
+    const auto inferred_val = expr_eval.evaluate(size);
     TORCH_INTERNAL_ASSERT(
         inferred_val.has_value(),
         "Could not launch kernel as program could not infer ",
-        kir::toString(id->rawExtent()),
+        kir::toString(size),
         " for the buffer ",
         kir::toString(tv));
-    sizes.push_back(inferred_val.value());
+    inferred_sizes.push_back(inferred_val.value());
   }
 
   const auto at_type = data_type_to_aten(tv->dtype());
 
   if (zero_init) {
     const auto tensor_options =
         at::TensorOptions().dtype(at_type).device(options.device);
-    c10::IntArrayRef isizes(sizes);
+    c10::IntArrayRef isizes(inferred_sizes);
     return at::zeros(isizes, tensor_options);
   } else {
-    c10::IntArrayRef isizes(sizes);
+    c10::IntArrayRef isizes(inferred_sizes);
     // Non Variable type guard for empty_cuda call
     at::AutoNonVariableTypeMode non_variable_type_mode;
     return at::native::empty_cuda(
         isizes, at_type, c10::nullopt, options.device, c10::nullopt);
   }
 }
 
+at::Tensor inferAndAllocOutput(
```

Comment:
It seems today the only thing we use allocation node for during execution is for shared memory buffers. Seems like we should be looking at allocation nodes directly and not inferring sizes at runtime from tensor views.

Reply:
I actually tried to get rid of this function as well and created a `kir::Allocaate` for output tensors as well, but that didn't quite feel natural to me. Where should allocation nodes be placed in a kernel IR? Allocation of outputs are done before a kernel is executed, so it seems odd to have allocation nodes inside a kernel. 

Having said that, I just realized that's also the case for dynamic shared memory. We place allocation nodes for dynamic shared memory just at the beginning of a kernel. Does that make sense for global memory?

Reply:
I wonder if we should just have a list of allocate nodes that aren't registers attached to the kernel itself. Instead of trying to place them inline. I don't recall though how aliasing works and if it needs to be ordered in the code to work.

Reply:
Going to unblock this PR, but maybe we should try cleaning up a bit of technical debt here in a follow up.

Reply:
Looking at `reuseMemoryAllocations`, I don't think it relies on `Allocate` nodes being inlined. 

Created an issue for a follow-up (#823).